File: qlang-interpreter/src/qlang/runtime/callable.ts
import { FunctionStatement } from "../ast"
import Environment from "./environment"
import Interpreter from "./interpreter"
import { AlgebraicValue } from "./values"

export type CallableCallback = (interpreter: Interpreter, ...args: AlgebraicValue[]) => AlgebraicValue

export interface Callable {

    get Arity(): number

    call(interpreter: Interpreter, args: AlgebraicValue[]): AlgebraicValue

    toString(): string

}

export class QFunction implements Callable {

    public static counter: number = 0

    private readonly func: FunctionStatement
    private readonly closure: Environment
    private readonly name: string

    constructor(
        func: FunctionStatement,
        closure: Environment
    ) {
        this.func = func
        this.closure = closure

        if (func.identifier === null) {
            QFunction.counter++
            this.name = `anon#${QFunction.counter}`
        } else {
            this.name = func.identifier
        }
    }

    get Arity(): number {
        return this.func.parameters.length
    }

    get Name(): string {
        return this.name
    }

    call(interpreter: Interpreter, args: AlgebraicValue[]): AlgebraicValue {
        const environment = new Environment(this.closure)

        for (let i = 0; i < this.Arity; i++) {
            environment.declareVariable(this.func.parameters[i], args[i])
        }

        return interpreter.evaluateBlockStatement(this.func.body, environment) as AlgebraicValue
    }

    toString() {
        return `<fonction #${this.name}>`
    }
}

File: interpreter/src/runtime/callable.py
from abc import ABC
from abc import abstractmethod
from typing import TYPE_CHECKING
from typing import Self


if TYPE_CHECKING:
    from src.interpreter import Interpreter

from src.nodes.merged import FunctionStatement
from src.runtime.environment import Environment
from src.runtime.values import MK_NULL
from src.runtime.values import AlgebraicValue


class Callable(ABC):

    @abstractmethod
    def arity(self: Self) -> int:
        raise NotImplementedError("'arity' method must be implemented")

    @abstractmethod
    def call(self: Self, interpreter, args: list[AlgebraicValue]) -> AlgebraicValue:
        raise NotImplementedError("'call' method must be implemented")

    @abstractmethod
    def __str__(self: Self) -> str:
        raise NotImplementedError("'__str__' method must be implemented")


class QFunction(Callable):

    counter: int = 0

    def __init__(self, func: FunctionStatement, closure: Environment):
        self.func = func
        self.closure = closure

        if self.func.identifier is None:
            QFunction.counter += 1
            self.name = f"anon_{QFunction.counter}"
        else:
            self.name = self.func.identifier

    def arity(self: Self) -> int:
        return len(self.func.parameters)

    def call(
        self: Self, interpreter: "Interpreter", args: list[AlgebraicValue]
    ) -> AlgebraicValue:
        env = Environment(self.closure)

        for i, param in enumerate(self.func.parameters):
            env.declareVariable(param, args[i])

        value = interpreter.evaluate_block_statement(self.func.body, env)
        if value is AlgebraicValue:
            return value  # type: ignore

        return MK_NULL()

    def __str__(self: Self) -> str:
        return f"<fonction @{self.name}>"
